internal/core/src/cachinglayer/lrucache/DList.cpp

```diff
@@ -0,0 +1,201 @@
+// Copyright (C) 2019-2025 Zilliz. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under the License
+#include "cachinglayer/lrucache/DList.h"
+
+#include <mutex>
+#include <vector>
+
+#include <folly/futures/Future.h>
+#include <folly/futures/SharedPromise.h>
+
+#include "cachinglayer/Utils.h"
+#include "cachinglayer/lrucache/ListNode.h"
+#include "monitor/prometheus_client.h"
+
+namespace milvus::cachinglayer::internal {
+
+bool
+DList::reserveMemory(const ResourceUsage& size) {
+    std::unique_lock<std::mutex> list_lock(list_mtx_);
+    auto used = used_memory_.load();
+    if (max_memory_.CanHold(used + size)) {
+        used_memory_ += size;
+        return true;
+    }
+    if (tryEvict(used + size - max_memory_)) {
+        used_memory_ += size;
+        return true;
+    }
+    return false;
+}
+
+bool
+DList::tryEvict(const ResourceUsage& expected_eviction) {
```

Comment:
try to evict use a background thread from low water mark instead of evict on read. This may affect the performance and cause long stalls.

We should also have multiple eviction policies such as LRU, LFU and time based.

Reply:
also when we evict, usually we evict a lot be more than necessary.
otherwise we end into a state where each cache write will trigger a eviction

Reply:
> evict use a background thread from low water mark

Ack. Will update that in https://github.com/milvus-io/milvus/pull/41546. We have multiple config changes in that PR, and this will require config change, so will do it there.

> We should also have multiple eviction policies such as LRU, LFU and time based.

currently we have LRU as the default policy. how about we add LFU and time based policies in 2.6.x in the future?

> when we evict, usually we evict a lot be more than necessary

once we move to low/high water mark this should be the default behavior: once resource usage reached high water mark, we evict until it is lower than low water mark.